Security updates

jbrophy17 opened the pull request:
Modified the JWTAlgorithm protocol to add a required function, verifySignedInput:withSignature:verificationKey, to allow the algorithm to provide it's own verification logic. 

Changed the RS256 algorithm to no longer subclass the None algorithm, it isn't implemented, so it should fail verification (unless force decode is enabled)

Also a few changes to make decoding JWTs more secure

1) Decoding a JWT requires you to specify the algorithm you want to use. Trusting the header's value of "alg" is not safe since the JWT hasn't been verified yet. You don't want to allow an attacker to choose the algorithm to verify their JWT with.
   - The decoding function will now error if the 'withForcedAlgorithmByName' parameter is not set. This can be skipped using the forcedOption still, but should only be done for debugging purposes.
   - Any decode functions that didn't explicitly take in an algorithm name use HS512 by default (which matches the default algorithm used by the encode functions). I believe these methods should be deprecated since you should always specify an algorithm

2) If the specified algorithm and the algorithm value provided by the header don't match, verification fails. You should always know what algorithm you're using, and if you receive a JWT with an unexpected "alg" value, something is fishy

3) Added an optional whitelist for algorithms. If enabled, decoding will fail for any algorithm that hasn't been added to the whitelist

lolgear commented:
@jbrophy17, necessary features added. but I would ask about additional development as a part of version 2.0.0. it could be useful to deprecate all plain functions and only leave alone fluent part. also, it would be great to change whitelist to something more dynamic, not plain static fields.
thanks

jbrophy17 commented:
@lolgear Something like this? Deprecated all the Class encode/decode functions on JWT and added private encode/decode helper functions to the JWTBuilder object. Also changed the whitelist to be a property of a JWTBuilder and removed the static implementation

lolgear commented:
@jbrophy17, great job!
